File: Site/static/Site/js/pages/config/place/table/prepInfo.js
import {info} from "../../../../storage/config/place/info";
import {countriesList} from "../../../../storage/app/countriesList";
import {regionsList} from "../../../../storage/app/regionsList";
import {citiesList} from "../../../../storage/app/citiesList";

export const prepInfo = () => {
    info.clear();

    const regionPrefix = countriesList.value.length + 1000;
    const cityPrefix = regionPrefix + regionsList.value.length + 1000;

    countriesList.value.forEach((el, index) => {
        if (el.title !== '') {
            info.push({
                ...emptyRow,
                id: index,
                typeRow: 'country',
                realId: el.id,
                country: el.title
            });
        }
    });

    regionsList.value.forEach((el, index) => {
        const country = countriesList.value.find(elem => elem.id === el.country_id);
        if (el.title !== '' && country) {
            info.push({
                ...emptyRow,
                id: regionPrefix + index,
                typeRow: 'region',
                realId: el.id,
                country: country.title,
                country_id: country.id,
                region: el.title
            });
        }
    });

    citiesList.value.forEach((el, index) => {
        const country = countriesList.value.find(elem => elem.id === el.country_id);
        const region = regionsList.value.find(elem => elem.id === el.region_id);
        if (el.title !== '' && country) {
            info.push({
                ...emptyRow,
                id: cityPrefix + index,
                typeRow: 'city',
                realId: el.id,
                country: country.title,
                country_id: country.id,
                region: region?.title || '',
                region_id: region?.id || null,
                city: el.title
            });
        }
    })
};


const emptyRow = {
    id: '',
    country: '',
    region: '',
    city: ''
};

File: Site/controllers/place/prepInfo.py
def prepInfo(countriesList, regionsList, citiesList):
    emptyRow = {'id': '', 'country': '', 'region': '', 'city': ''}
    index = 0
    result = []

    for country in countriesList.iterator():
        info = {}
        info.update(emptyRow)
        info.update({'id': index, 'typeRow': 'country', 'realId': country.id, 'country': country.title})
        result.append(info)
        index += 1

    index += 1000

    for region in regionsList.iterator():
        info = {}
        info.update(emptyRow)
        info.update({'id': index, 'typeRow': 'region', 'realId': region.id, 'region': region.title,
                     'country': region.country.title, 'country_id': region.country.id})

        result.append(info)
        index += 1

    index += 1000

    for city in citiesList.iterator():
        info = {}
        info.update(emptyRow)
        info.update({'id': index, 'typeRow': 'city', 'realId': city.id, 'city': city.title,
                     'country': city.country.title, 'country_id': city.country.id,
                     'region': city.region.title if city.region else '',
                     'region_id': city.region.id if city.region else ''})

        result.append(info)
        index += 1

    return result
